File: mobile/ios/SthiraApp/Views/Components.swift
import SwiftUI

/// Banner identifying synthetic drill exercise data to prevent citizen panic.
struct SyntheticExerciseBanner: View {
    let isExercise: BooleanLiteralType

    var body: some View {
        if isExercise {
            HStack {
                Spacer()
                Text("⚠️ SYNTHETIC EXERCISE — NOT A REAL DISASTER ORDER")
                    .font(.system(size: 12, weight: .bold))
                    .foregroundColor(.white)
                    .multilineTextAlignment(.center)
                Spacer()
            }
            .padding(.vertical, 8)
            .padding(.horizontal, 16)
            .background(SthiraTheme.syntheticDemoBanner)
            .accessibilityElement(children: .combine)
            .accessibilityLabel("Notice: Synthetic disaster drill exercise in progress. This is not a real evacuation order.")
        }
    }
}

/// Displays data freshness status honestly (FRESH, STALE, UNVERIFIABLE, EXPIRED).
struct FreshnessBadge: View {
    let freshness: String
    let sourceId: String

    var body: some View {
        let (bgColor, textColor, label) = config(for: freshness)

        HStack(spacing: 4) {
            Text("\(label) • \(sourceId)")
                .font(.system(size: 11, weight: .semibold))
                .foregroundColor(textColor)
        }
        .padding(.horizontal, 8)
        .padding(.vertical, 4)
        .background(bgColor)
        .overlay(
            RoundedRectangle(cornerRadius: 4)
                .stroke(textColor, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: 4))
        .accessibilityLabel("Data status: \(label) from source \(sourceId)")
    }

    private func config(for freshness: String) -> (Color, Color, String) {
        switch freshness {
        case "FRESH":
            return (SthiraTheme.safeGreenSurface, SthiraTheme.safeGreen, "Verified Live")
        case "STALE":
            return (SthiraTheme.warningAmberSurface, SthiraTheme.warningAmber, "Stale Data (Needs Recheck)")
        case "UNVERIFIABLE":
            return (SthiraTheme.hazardRedSurface, SthiraTheme.hazardRed, "Unverifiable (Clock Mismatch)")
        case "EXPIRED":
            return (SthiraTheme.hazardRedSurface, SthiraTheme.hazardRed, "Expired Advisory")
        default:
            return (Color(white: 0.9), Color(white: 0.4), "Withdrawn")
        }
    }
}

/// Prominent card offering instant emergency call to 112 / local disaster control room.
struct EmergencyCallCard: View {
    let officialNumber: String

    init(officialNumber: String = "112") {
        self.officialNumber = officialNumber
    }

    var body: some View {
        HStack {
            VStack(alignment: .leading, spacing: 2) {
                Text("Immediate Danger?")
                    .font(.system(size: 16, weight: .bold))
                    .foregroundColor(SthiraTheme.hazardRed)

                Text("Contact official disaster emergency services")
                    .font(.system(size: 12))
                    .foregroundColor(SthiraTheme.textSecondaryLight)
            }

            Spacer()

            Button(action: {
                EmergencyDialler.call112(number: officialNumber)
            }) {
                Text("Call \(officialNumber)")
                    .font(.system(size: 14, weight: .bold))
                    .foregroundColor(.white)
                    .padding(.horizontal, 14)
                    .padding(.vertical, 10)
                    .background(SthiraTheme.hazardRed)
                    .cornerRadius(8)
            }
            .accessibleTarget(minPt: 44)
            .accessibilityLabel("Call emergency services at \(officialNumber)")
        }
        .padding(14)
        .background(SthiraTheme.hazardRedSurface)
        .cornerRadius(10)
    }
}

/// Displays facility bed capacity truthfully without guessing.
struct CapacityIndicator: View {
    let remainingCapacity: Int?
    let totalCapacity: Int?

    var body: some View {
        if let rem = remainingCapacity, let tot = totalCapacity {
            if rem <= 0 {
                Text("⛔ Full (0/\(tot) Available)")
                    .font(.system(size: 12, weight: .bold))
                    .foregroundColor(SthiraTheme.hazardRed)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(SthiraTheme.hazardRedSurface)
                    .cornerRadius(4)
                    .accessibilityLabel("Shelter at maximum capacity. 0 beds available.")
            } else {
                Text("✓ \(rem)/\(tot) Available")
                    .font(.system(size: 12, weight: .bold))
                    .foregroundColor(SthiraTheme.safeGreen)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(SthiraTheme.safeGreenSurface)
                    .cornerRadius(4)
                    .accessibilityLabel("\(rem) of \(tot) beds available")
            }
        } else {
            Text("❓ Capacity: Unconfirmed")
                .font(.system(size: 12, weight: .semibold))
                .foregroundColor(Color(red: 81/255, green: 45/255, blue: 168/255))
                .padding(.horizontal, 8)
                .padding(.vertical, 4)
                .background(Color(red: 237/255, green: 231/255, blue: 246/255))
                .cornerRadius(4)
                .accessibilityLabel("Capacity status unconfirmed by authority")
        }
    }
}

File: mobile/android/src/main/kotlin/org/sthira/mobile/android/ui/Components.kt
package org.sthira.mobile.android.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import org.sthira.mobile.android.accessibility.EmergencyDialler
import org.sthira.mobile.android.accessibility.accessibleTarget
import org.sthira.mobile.android.theme.HazardRed
import org.sthira.mobile.android.theme.SafeGreen
import org.sthira.mobile.android.theme.SthiraBluePrimary
import org.sthira.mobile.android.theme.SyntheticDemoBanner
import org.sthira.mobile.android.theme.WarningAmber
import org.sthira.mobile.contracts.FreshnessState

/**
 * SyntheticExerciseBanner: clearly labels exercise data to prevent citizens
 * from mistaking drills for real disaster evacuations.
 */
@Composable
fun SyntheticExerciseBanner(
    isExercise: Boolean,
    modifier: Modifier = Modifier
) {
    if (!isExercise) return

    Box(
        modifier = modifier
            .fillMaxWidth()
            .background(SyntheticDemoBanner)
            .padding(horizontal = 16.dp, vertical = 8.dp)
            .semantics {
                contentDescription = "Notice: Synthetic disaster drill exercise in progress. This is not a real evacuation order."
            },
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = "⚠️ SYNTHETIC EXERCISE — NOT A REAL DISASTER ORDER",
            color = Color.White,
            fontWeight = FontWeight.Bold,
            fontSize = 12.sp,
            letterSpacing = 0.5.sp
        )
    }
}

/**
 * FreshnessBadge: displays honest freshness states (FRESH, STALE, UNVERIFIABLE, EXPIRED)
 */
@Composable
fun FreshnessBadge(
    freshness: FreshnessState,
    sourceId: String,
    modifier: Modifier = Modifier
) {
    val (bgColor, textColor, label) = when (freshness) {
        FreshnessState.FRESH -> Triple(Color(0xFFE8F5E9), SafeGreen, "Verified Live")
        FreshnessState.STALE -> Triple(Color(0xFFFFF3E0), WarningAmber, "Stale Data (Needs Recheck)")
        FreshnessState.UNVERIFIABLE -> Triple(Color(0xFFFFEBEE), HazardRed, "Unverifiable (Clock Mismatch)")
        FreshnessState.EXPIRED -> Triple(Color(0xFFFFEBEE), HazardRed, "Expired Advisory")
        FreshnessState.WITHDRAWN -> Triple(Color(0xFFEEEEEE), Color(0xFF616161), "Withdrawn by Authority")
    }

    Box(
        modifier = modifier
            .background(bgColor, shape = RoundedCornerShape(4.dp))
            .border(1.dp, textColor, shape = RoundedCornerShape(4.dp))
            .padding(horizontal = 8.dp, vertical = 4.dp)
            .semantics { contentDescription = "Data status: $label from $sourceId" }
    ) {
        Text(
            text = "$label • $sourceId",
            color = textColor,
            fontSize = 11.sp,
            fontWeight = FontWeight.SemiBold
        )
    }
}

/**
 * EmergencyCallCard: provides a prominent button to open the device dialler
 * with 112 / local disaster helpline.
 */
@Composable
fun EmergencyCallCard(
    officialContact: String = "112",
    modifier: Modifier = Modifier
) {
    val context = LocalContext.current

    Card(
        modifier = modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(containerColor = Color(0xFFFFEBEE)),
        shape = RoundedCornerShape(8.dp)
    ) {
        Row(
            modifier = Modifier
                .padding(16.dp)
                .fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Column(modifier = Modifier.weight(1f)) {
                Text(
                    text = "Immediate Danger?",
                    fontWeight = FontWeight.Bold,
                    fontSize = 16.sp,
                    color = HazardRed
                )
                Text(
                    text = "Contact official emergency services",
                    fontSize = 13.sp,
                    color = Color(0xFF4A4A4A)
                )
            }
            Spacer(modifier = Modifier.width(12.dp))
            Button(
                onClick = { EmergencyDialler.openDialler(context, officialContact) },
                colors = ButtonDefaults.buttonColors(containerColor = HazardRed),
                modifier = Modifier.accessibleTarget(48)
            ) {
                Text(
                    text = "Call $officialContact",
                    color = Color.White,
                    fontWeight = FontWeight.Bold
                )
            }
        }
    }
}

/**
 * CapacityIndicator: displays shelter availability honestly.
 * Unknown capacity is explicitly presented as unknown, never 0 or 100%.
 */
@Composable
fun CapacityIndicator(
    remainingCapacity: Int?,
    totalCapacity: Int?,
    modifier: Modifier = Modifier
) {
    when {
        remainingCapacity == null || totalCapacity == null -> {
            Box(
                modifier = modifier
                    .background(Color(0xFFEDE7F6), RoundedCornerShape(4.dp))
                    .padding(horizontal = 8.dp, vertical = 4.dp)
                    .semantics { contentDescription = "Capacity unconfirmed by local authority" }
            ) {
                Text(
                    text = "❓ Capacity: Unconfirmed",
                    color = Color(0xFF4A148C),
                    fontSize = 12.sp,
                    fontWeight = FontWeight.SemiBold
                )
            }
        }
        remainingCapacity <= 0 -> {
            Box(
                modifier = modifier
                    .background(Color(0xFFFFEBEE), RoundedCornerShape(4.dp))
                    .padding(horizontal = 8.dp, vertical = 4.dp)
                    .semantics { contentDescription = "Shelter at maximum capacity. 0 beds available." }
            ) {
                Text(
                    text = "⛔ Full (0/$totalCapacity Available)",
                    color = HazardRed,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold
                )
            }
        }
        else -> {
            Box(
                modifier = modifier
                    .background(Color(0xFFE8F5E9), RoundedCornerShape(4.dp))
                    .padding(horizontal = 8.dp, vertical = 4.dp)
                    .semantics { contentDescription = "$remainingCapacity of $totalCapacity beds available" }
            ) {
                Text(
                    text = "✓ $remainingCapacity/$totalCapacity Available",
                    color = SafeGreen,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold
                )
            }
        }
    }
}
